File: Universidad de Granada/Programación y Diseño Orientada a Objetos/QytetetJava/src/modeloqytetet/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author Miguel Ángel Benítez Alguacil
 *         Najib Saadouli Arco
 */
public class Especulador extends Jugador {
    private int fianza;
    
    protected Especulador(Jugador jugador, int fianza){
        super(jugador);
        this.fianza = fianza;
    }
    
    @Override
    protected void pagarImpuesto(){
        modificarSaldo(-getCasillaActual().getCoste()/2);
    }
    
    @Override
    protected Especulador convertirme(int fianza){
        Especulador especulador = new Especulador(this, fianza);
        return especulador;
    }
    
    @Override
    protected boolean deboIrACarcel (){
        return ( super.deboIrACarcel() && !this.pagarFianza() );
    }
    
    private boolean pagarFianza(){
        if(tengoSaldo(fianza)){
            this.modificarSaldo(-fianza);
            return true;
        }
        else
            return false;
    }
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad titulo){
        return ( tengoSaldo(titulo.getPrecioEdificar()) && titulo.getNumCasas()<8 );
    }
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad titulo){
        return ( tengoSaldo(titulo.getPrecioEdificar()) && titulo.getNumCasas()>=4 && titulo.getNumHoteles()<8 );
    }

    @Override
    public String toString() {
        return "\n\tEspeculador{" + "nombre = " + getNombre() + ", encarcelado = " + getEncarcelado()
                + ", saldo = " + getSaldo() + ", carta libertad = " + getCartaLibertad()
                + ", casilla actual = " + getCasillaActual() + ", propiedades = " + getPropiedades()
                + ", capital del jugador = " + obtenerCapital() + ", fianza=" + fianza + "}";
    }
    
    
}

File: Universidad de Granada/Programación y Diseño Orientada a Objetos/QytetetRuby/lib/especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

#authors: Miguel Ángel Benítez Alguacil
#         Najib Saadouli Arco

require_relative "jugador"

module ModeloQytetet
  class Especulador < Jugador
    def self.copia(otroJugador, fianza)
      especulador = super(otroJugador)
      especulador.fianza = fianza
      
      especulador
    end
    
    attr_accessor :fianza
    
    def pagarImpuesto
      modificarSaldo(-@casillaActual.coste/2)
    end
    
    def convertirme(fianza)
      Especulador.copia(self, fianza)
    end
    
    def deboIrACarcel
      super && !pagarFianza
    end
    
    def pagarFianza
      if(tengoSaldo(@fianza))
        modificarSaldo(-@fianza)
        true
      else
        false
      end
    end
    
    def puedoEdificarCasa(titulo)
      tengoSaldo(titulo.precioEdificar) && titulo.numCasas < 8
    end
    
    def puedoEdificarHotel(titulo)
      tengoSaldo(titulo.precioEdificar) && titulo.numCasas >= 4 && titulo.numHoteles < 8
    end
    
    private :pagarFianza
    protected :puedoEdificarCasa, :puedoEdificarHotel
    
    def to_s
      retorno = "\n\tEspeculador {nombre = #{@nombre}, saldo = #{@saldo}, encarcelado = #{@encarcelado}, cartaLibertad = #{@cartaLibertad}, casillaActual = #{@casillaActual}, propiedades = "
      @propiedades.each do |propiedad|    # saldo = #{obtenerCapital},
        retorno << propiedad.to_s 
      end
      retorno << "\n\tcapital del jugador = #{obtenerCapital}, fianza = #{@fianza}}\n"
    end
  end
end
